pytorch_lightning/callbacks/quantization.py

```diff
@@ -0,0 +1,201 @@
+        self._qconfig = qconfig
+
+        if observer_type not in self.OBSERVER_TYPES:
+            raise MisconfigurationException(
+                f'Unsupported observer type "{observer_type}", allowed are {self.OBSERVER_TYPES}.'
+            )
+        self._observer_type = observer_type
+
+        if collect_quantization is not None and not isinstance(collect_quantization, (int, Callable)):
+            raise MisconfigurationException(
+                f'Unsupported `collect_quantization` "{observer_type}", allowed are int or callable.'
+            )
+        self._collect_quantization = collect_quantization
+
+        self.modules_to_fuse = modules_to_fuse
+        self._preserve_compatible = input_compatible
+        self._forward_calls = 0
+
+    def _check_feasible_fuse(self, model):
+        if not self.modules_to_fuse:
+            return True
+        for group in self.modules_to_fuse:
+            if not all(_recursive_hasattr(model, m) for m in group):
+                raise MisconfigurationException(
+                    f'You have requested to fuse {group} but one or more of them is not your model attributes'
+                )
+        return True
+
+    def on_fit_start(self, trainer, pl_module):
+        # QuantStub converts tensors from floating point to quantized
+        pl_module.quant = torch.quantization.QuantStub()
+        # DeQuantStub converts tensors from quantized to floating point
+        pl_module.dequant = torch.quantization.DeQuantStub()
+        # manually specify where tensors will be converted from quantized
+        # to floating point in the quantized model
+        self.__module_forward = pl_module.forward
+        pl_module.forward = wrap_qat_forward_context(
+            quant_cb=self, model=pl_module, func=pl_module.forward, trigger_condition=self._collect_quantization
+        )
+
+        # attach a global qconfig, which contains information about what kind
+        # of observers to attach. Use 'fbgemm' for server inference
+        if isinstance(self._qconfig, str):
+            if self._observer_type == 'histogram':
+                pl_module.qconfig = torch.quantization.get_default_qconfig(self._qconfig)
+            elif self._observer_type == 'average':
+                pl_module.qconfig = torch.quantization.get_default_qat_qconfig(self._qconfig)
+        elif isinstance(self._qconfig, QConfig):
+            pl_module.qconfig = self._qconfig
+
+        if self.modules_to_fuse:
+            self._check_feasible_fuse(pl_module)
+            torch.quantization.fuse_modules(pl_module, self.modules_to_fuse, inplace=True)
+
+        # Prepare the model for QAT. This inserts observers and fake_quants in
+        # the model that will observe weight and activation tensors during calibration.
+        torch.quantization.prepare_qat(pl_module, inplace=True)
+
+    def on_fit_end(self, trainer, pl_module):
+        pl_module.eval()
```

Comment:
why do we set eval here? Can't the user call `.eval()` after the model is converted?

Reply:
from docs, it seems that you have to have a model in eval mode to perform conversion... (I may try it without)